File: sync-firebase-to-mongodb.js
#!/usr/bin/env node
/**
 * Firebase to MongoDB Data Sync Script
 * Restores all Firebase data to MongoDB backup
 */

const admin = require('firebase-admin');
const { MongoClient } = require('mongodb');
const fs = require('fs');
const path = require('path');
require('dotenv').config({ path: '.env.local' });

// Firebase configuration
const firebaseConfig = {
  apiKey: process.env.VITE_FIREBASE_API_KEY,
  authDomain: process.env.VITE_FIREBASE_AUTH_DOMAIN,
  projectId: process.env.VITE_FIREBASE_PROJECT_ID,
  storageBucket: process.env.VITE_FIREBASE_STORAGE_BUCKET,
  messagingSenderId: process.env.VITE_FIREBASE_MESSAGING_SENDER_ID,
  appId: process.env.VITE_FIREBASE_APP_ID,
};

// MongoDB configuration
const mongoUri = process.env.MONGODB_URI;
const mongoDbName = process.env.MONGODB_DB_NAME;

let db_firestore = null;
let db_mongo = null;
let mongoClient = null;

async function initializeFirebase() {
  try {
    // Check for service account key
    const serviceAccountPath = path.join(__dirname, 'firebase-service-account.json');
    
    if (fs.existsSync(serviceAccountPath)) {
      const serviceAccount = require('./firebase-service-account.json');
      admin.initializeApp({
        credential: admin.credential.cert(serviceAccount),
        projectId: firebaseConfig.projectId,
      });
      console.log('[✓] Firebase initialized with service account key');
    } else {
      console.log('[!] Service account key not found');
      console.log('    You can download it from:');
      console.log('    Firebase Console > Project Settings > Service Accounts');
      console.log('    Place it as: firebase-service-account.json');
      return false;
    }
    
    db_firestore = admin.firestore();
    return true;
  } catch (error) {
    console.error('[!] Firebase initialization failed:', error.message);
    return false;
  }
}

async function connectMongoDB() {
  try {
    mongoClient = new MongoClient(mongoUri);
    await mongoClient.connect();
    db_mongo = mongoClient.db(mongoDbName);
    console.log('[✓] Connected to MongoDB');
    return true;
  } catch (error) {
    console.error('[!] MongoDB connection failed:', error.message);
    return false;
  }
}

async function syncMessages() {
  if (!db_firestore || !db_mongo) return 0;
  
  try {
    console.log('[*] Syncing chat messages from Firebase...');
    
    const appId = 'access-ai-v4';
    const usersRef = db_firestore.collection('artifacts').doc(appId).collection('users');
    
    let totalSynced = 0;
    const usersSnapshot = await usersRef.get();
    
    for (const userDoc of usersSnapshot.docs) {
      const userId = userDoc.id;
      console.log(`  [*] Processing user: ${userId}`);
      
      const chatsRef = userDoc.ref.collection('chats');
      const chatsSnapshot = await chatsRef.get();
      
      for (const chatDoc of chatsSnapshot.docs) {
        const chatId = chatDoc.id;
        const messagesRef = chatDoc.ref.collection('messages');
        const messagesSnapshot = await messagesRef.get();
        
        for (const msgDoc of messagesSnapshot.docs) {
          const msgData = msgDoc.data();
          
          try {
            // Add metadata for tracking
            await db_mongo.collection('messages').updateOne(
              { firebaseId: msgDoc.id, userId },
              {
                $set: {
                  userId,
                  chatId,
                  firebaseId: msgDoc.id,
                  ...msgData,
                  syncedAt: new Date(),
                }
              },
              { upsert: true }
            );
            totalSynced++;
          } catch (error) {
            if (!error.message.includes('duplicate')) {
              console.error(`    [!] Error syncing message: ${error.message}`);
            }
          }
        }
      }
    }
    
    console.log(`[✓] Synced ${totalSynced} messages to MongoDB`);
    return totalSynced;
  } catch (error) {
    console.error('[!] Error syncing messages:', error);
    return 0;
  }
}

async function syncUserProfiles() {
  if (!db_firestore || !db_mongo) return 0;
  
  try {
    console.log('[*] Syncing user profiles from Firebase...');
    
    const appId = 'access-ai-v4';
    const usersRef = db_firestore.collection('artifacts').doc(appId).collection('users');
    
    let totalSynced = 0;
    const usersSnapshot = await usersRef.get();
    
    for (const userDoc of usersSnapshot.docs) {
      const userId = userDoc.id;
      const userData = userDoc.data();
      
      try {
        await db_mongo.collection('users').updateOne(
          { uid: userId },
          {
            $set: {
              uid: userId,
              firebaseId: userDoc.id,
              ...userData,
              syncedAt: new Date(),
            }
          },
          { upsert: true }
        );
        totalSynced++;
      } catch (error) {
        console.error(`  [!] Error syncing user ${userId}: ${error.message}`);
      }
    }
    
    console.log(`[✓] Synced ${totalSynced} user profiles to MongoDB`);
    return totalSynced;
  } catch (error) {
    console.error('[!] Error syncing user profiles:', error);
    return 0;
  }
}

async function syncEmergencyContacts() {
  if (!db_firestore || !db_mongo) return 0;
  
  try {
    console.log('[*] Syncing emergency contacts from Firebase...');
    
    const appId = 'access-ai-v4';
    const usersRef = db_firestore.collection('artifacts').doc(appId).collection('users');
    
    let totalSynced = 0;
    const usersSnapshot = await usersRef.get();
    
    for (const userDoc of usersSnapshot.docs) {
      const userId = userDoc.id;
      
      try {
        const contactsRef = userDoc.ref.collection('emergencyContacts');
        const contactsSnapshot = await contactsRef.get();
        
        for (const contactDoc of contactsSnapshot.docs) {
          const contactData = contactDoc.data();
          
          try {
            await db_mongo.collection('emergency_contacts').updateOne(
              { firebaseId: contactDoc.id, userId },
              {
                $set: {
                  userId,
                  firebaseId: contactDoc.id,
                  ...contactData,
                  syncedAt: new Date(),
                }
              },
              { upsert: true }
            );
            totalSynced++;
          } catch (error) {
            if (!error.message.includes('duplicate')) {
              console.error(`    [!] Error syncing contact: ${error.message}`);
            }
          }
        }
      } catch (error) {
        // No emergency contacts for this user
      }
    }
    
    console.log(`[✓] Synced ${totalSynced} emergency contacts to MongoDB`);
    return totalSynced;
  } catch (error) {
    console.error('[!] Error syncing emergency contacts:', error);
    return 0;
  }
}

async function performFullSync() {
  console.log('='.repeat(70));
  console.log('FIREBASE TO MONGODB DATA SYNC');
  console.log('='.repeat(70));
  
  // Check prerequisites
  if (!mongoUri) {
    console.error('[!] MONGODB_URI not set in .env.local');
    return false;
  }
  
  if (!process.env.VITE_FIREBASE_API_KEY) {
    console.error('[!] Firebase configuration missing from .env.local');
    console.error('    Required: VITE_FIREBASE_API_KEY, VITE_FIREBASE_PROJECT_ID, etc.');
    return false;
  }
  
  // Initialize connections
  const firebaseOK = await initializeFirebase();
  if (!firebaseOK) {
    console.error('[!] Firebase initialization failed');
    return false;
  }
  
  const mongoOK = await connectMongoDB();
  if (!mongoOK) {
    console.error('[!] MongoDB connection failed');
    return false;
  }
  
  console.log('[✓] Connected to both Firebase and MongoDB');
  console.log();
  
  // Sync all data
  const msgCount = await syncMessages();
  const userCount = await syncUserProfiles();
  const contactCount = await syncEmergencyContacts();
  
  console.log();
  console.log('='.repeat(70));
  console.log('SYNC COMPLETE!');
  console.log('='.repeat(70));
  console.log(`✅ Messages synced: ${msgCount}`);
  console.log(`✅ Users synced: ${userCount}`);
  console.log(`✅ Emergency contacts synced: ${contactCount}`);
  console.log(`✅ Total records: ${msgCount + userCount + contactCount}`);
  console.log();
  console.log('All Firebase data has been backed up to MongoDB!');
  console.log('='.repeat(70));
  
  return true;
}

// Main execution
(async () => {
  try {
    const success = await performFullSync();
    process.exit(success ? 0 : 1);
  } catch (error) {
    console.error('[FATAL]', error);
    process.exit(1);
  } finally {
    if (mongoClient) {
      await mongoClient.close();
    }
    if (admin.apps.length > 0) {
      await admin.app().delete();
    }
  }
})();

File: sync_firebase_to_mongodb.py
#!/usr/bin/env python3
# -*- coding: utf-8 -*-
"""
Firebase to MongoDB Data Sync
Restores all existing Firebase data to MongoDB as backup
"""

import os
import json
from pathlib import Path
from dotenv import load_dotenv

# Load environment variables
try:
    env_path = Path(__file__).parent / '.env.local'
    if env_path.exists():
        load_dotenv(env_path)
except ImportError:
    pass

try:
    import firebase_admin
    from firebase_admin import credentials, firestore
    FIREBASE_AVAILABLE = True
except ImportError:
    FIREBASE_AVAILABLE = False
    print("[!] Firebase admin SDK not available")

try:
    from mongodb_connection import get_db, insert_message, save_user_profile, save_emergency_contact
    MONGODB_AVAILABLE = True
except ImportError:
    MONGODB_AVAILABLE = False
    print("[!] MongoDB module not available")

def init_firebase():
    """Initialize Firebase Admin SDK"""
    try:
        # Try to use service account key if available
        service_account_path = Path(__file__).parent / 'firebase-service-account.json'
        
        if service_account_path.exists():
            cred = credentials.Certificate(str(service_account_path))
            firebase_admin.initialize_app(cred)
            print("[✓] Firebase initialized with service account")
            return True
        else:
            print("[!] Firebase service account key not found at:", service_account_path)
            print("[!] Please download it from Firebase Console > Project Settings > Service Accounts")
            return False
    except Exception as e:
        print(f"[!] Firebase initialization error: {e}")
        return False

def sync_messages(db_firestore, db_mongo):
    """Sync all chat messages from Firebase to MongoDB"""
    try:
        print("[*] Syncing messages from Firebase...")
        
        # Get all messages collection reference
        messages_ref = db_firestore.collection('artifacts').document('access-ai-v4').collection('users')
        
        total_synced = 0
        
        # Iterate through all users
        for user_doc in messages_ref.stream():
            user_id = user_doc.id
            print(f"  [*] Processing user: {user_id}")
            
            try:
                # Get all chats for this user
                chats_ref = user_doc.reference.collection('chats')
                
                for chat_doc in chats_ref.stream():
                    chat_id = chat_doc.id
                    
                    # Get all messages for this chat
                    messages = chat_doc.reference.collection('messages').stream()
                    
                    for msg in messages:
                        msg_data = msg.to_dict()
                        
                        # Save to MongoDB
                        try:
                            db_mongo['messages'].insert_one({
                                'userId': user_id,
                                'chatId': chat_id,
                                'firebaseId': msg.id,
                                **msg_data
                            })
                            total_synced += 1
                        except Exception as e:
                            if 'duplicate' not in str(e).lower():
                                print(f"    [!] Error syncing message: {e}")
            
            except Exception as e:
                print(f"  [!] Error reading user {user_id}: {e}")
        
        print(f"[✓] Synced {total_synced} messages to MongoDB")
        return total_synced
        
    except Exception as e:
        print(f"[!] Error syncing messages: {e}")
        return 0

def sync_users(db_firestore, db_mongo):
    """Sync user profiles from Firebase to MongoDB"""
    try:
        print("[*] Syncing user profiles from Firebase...")
        
        users_ref = db_firestore.collection('artifacts').document('access-ai-v4').collection('users')
        
        total_synced = 0
        
        for user_doc in users_ref.stream():
            user_id = user_doc.id
            user_data = user_doc.to_dict()
            
            try:
                db_mongo['users'].update_one(
                    {'uid': user_id},
                    {'$set': {
                        'uid': user_id,
                        'firebaseId': user_doc.id,
                        **user_data
                    }},
                    upsert=True
                )
                total_synced += 1
            except Exception as e:
                print(f"  [!] Error syncing user {user_id}: {e}")
        
        print(f"[✓] Synced {total_synced} user profiles to MongoDB")
        return total_synced
        
    except Exception as e:
        print(f"[!] Error syncing users: {e}")
        return 0

def sync_emergency_contacts(db_firestore, db_mongo):
    """Sync emergency contacts from Firebase to MongoDB"""
    try:
        print("[*] Syncing emergency contacts from Firebase...")
        
        users_ref = db_firestore.collection('artifacts').document('access-ai-v4').collection('users')
        
        total_synced = 0
        
        for user_doc in users_ref.stream():
            user_id = user_doc.id
            
            try:
                # Get emergency contacts for this user
                contacts_ref = user_doc.reference.collection('emergencyContacts')
                
                for contact_doc in contacts_ref.stream():
                    contact_data = contact_doc.to_dict()
                    
                    try:
                        db_mongo['emergency_contacts'].insert_one({
                            'userId': user_id,
                            'firebaseId': contact_doc.id,
                            **contact_data
                        })
                        total_synced += 1
                    except Exception as e:
                        if 'duplicate' not in str(e).lower():
                            print(f"    [!] Error syncing contact: {e}")
            
            except Exception as e:
                # No emergency contacts collection for this user
                pass
        
        print(f"[✓] Synced {total_synced} emergency contacts to MongoDB")
        return total_synced
        
    except Exception as e:
        print(f"[!] Error syncing emergency contacts: {e}")
        return 0

def full_sync():
    """Execute full sync from Firebase to MongoDB"""
    
    if not FIREBASE_AVAILABLE:
        print("[!] Firebase SDK not available. Install: pip install firebase-admin")
        return False
    
    if not MONGODB_AVAILABLE:
        print("[!] MongoDB module not available")
        return False
    
    print("=" * 70)
    print("FIREBASE TO MONGODB DATA SYNC")
    print("=" * 70)
    
    # Initialize Firebase
    if not init_firebase():
        print("[!] Cannot initialize Firebase")
        return False
    
    # Get database connections
    try:
        db_firestore = firestore.client()
        db_mongo = get_db()
        
        if not db_mongo:
            print("[!] MongoDB connection failed")
            return False
        
        print("[✓] Connected to both Firebase and MongoDB")
        print()
        
        # Sync all data
        msg_count = sync_messages(db_firestore, db_mongo)
        user_count = sync_users(db_firestore, db_mongo)
        contact_count = sync_emergency_contacts(db_firestore, db_mongo)
        
        print()
        print("=" * 70)
        print("SYNC COMPLETE!")
        print("=" * 70)
        print(f"✅ Messages synced: {msg_count}")
        print(f"✅ Users synced: {user_count}")
        print(f"✅ Emergency contacts synced: {contact_count}")
        print(f"✅ Total records: {msg_count + user_count + contact_count}")
        print()
        print("All Firebase data has been backed up to MongoDB!")
        print("=" * 70)
        
        return True
        
    except Exception as e:
        print(f"[!] Sync error: {e}")
        import traceback
        print(traceback.format_exc())
        return False

if __name__ == '__main__':
    success = full_sync()
    exit(0 if success else 1)
